### ISSUE TITLE
[IMPROVEMENT] Crossreducer computes the mean over all edge types, not all received edge types.

### ISSUE THREAD

**Issue by fmellomascarenhas** [dmlc/dgl#3321]
## 🚀 Feature
in hg.multi_update_all, crossreducer should have the option to calculate the mean over only incoming edge types, and not all possible incoming edge types.

## Motivation

In the image below, every node has a message = 1. We are trying to take the mean message from yellow and blue nodes to the red node. 
One of the red nodes received 3 for yellow and 1 from blue, so the mean message should be 3+1/2 = 2.
The other red node received 2 from the blue node, so the message should be 2/1 = 2. However, this is not the case. crossreducer will calculate 2 blue + 0 yellow / 2 = 1.

![Screenshot from 2021-09-03 21-31-31](https://user-images.githubusercontent.com/23004953/132078080-8b1323d8-b703-47f2-b190-623dfc5ba0df.png)

```
import torch
import dgl
import dgl.function as fn

G = dgl.heterograph({('user', 'plays', 'game'): ([0, 1, 2], [0, 1, 1]),
                    ('store', 'sells', 'game'): ([0, 1, 2], [0, 0, 0])})

for ntype in G.ntypes:
    v = torch.ones(G.num_nodes(ntype), device = 'cpu').float().view(-1)
    G.nodes[ntype].data['v'] = v

#normalize incoming vector by the sum of scores
G.multi_update_all({etype : (fn.copy_u('v', 'v_normalized'), fn.sum('v_normalized', 'v_normalized')) \
                    for etype in  G.canonical_etypes}, cross_reducer = 'mean')

print(G.nodes['game'].data)
#>>> {'v': tensor([1., 1.]), 'v_normalized': tensor([2., 1.])}
```

If we add one more edgetype, we will see that now the crossreducer will calculate the (2 + 0 + 0)/3 = 0.667 (ignore the change in labels and colors. It is a bug in my plotting code)

![Screenshot from 2021-09-03 21-34-49](https://user-images.githubusercontent.com/23004953/132078164-0c49b039-96d1-4b8c-94fc-4e89a11b6577.png)

```
import torch
import dgl
import dgl.function as fn

G = dgl.heterograph({('user', 'plays', 'game'): ([0, 1, 2], [0, 1, 1]),
                    ('store', 'sells', 'game'): ([0, 1, 2], [0, 0, 0]),
                    ('company', 'makes', 'game'): ([0, 1, 2], [0, 0, 0])})

for ntype in G.ntypes:
    v = torch.ones(G.num_nodes(ntype), device = 'cpu').float().view(-1)
    G.nodes[ntype].data['v'] = v

#normalize incoming vector by the sum of scores
G.multi_update_all({etype : (fn.copy_u('v', 'v_normalized'), fn.sum('v_normalized', 'v_normalized')) \
                    for etype in  G.canonical_etypes}, cross_reducer = 'mean')

print(G.nodes['game'].data)
#>>{'v': tensor([1., 1.]), 'v_normalized': tensor([2.3333, 0.6667])}
```

**Comment by Rhett-Ying** [comment]
In your case, I think you could define your own cross_reducer like this:

`
def cross_mean(flist):
    sf = torch.stack(flist, dim=0)
    return torch.sum(sf, dim=0)/torch.count_nonzero(sf, dim=0)
`

Pls refine the above one in case of denominator is zero when dividing.

**Comment by fmellomascarenhas** [comment]
@Rhett-Ying , thanks for repplying! It works, but takes twice as much time. I wonder if you think it would make sense to support this kind of mean as a built in function.

```
import torch
import dgl
import dgl.function as fn

n_nodes = 100000
G = dgl.heterograph({('user', 'plays', 'game'): (torch.arange(n_nodes)[torch.randperm(n_nodes)], torch.arange(n_nodes)[torch.randperm(n_nodes)]),
                    ('store', 'sells', 'game'): (torch.arange(n_nodes)[torch.randperm(n_nodes)], torch.arange(n_nodes)[torch.randperm(n_nodes)]),
                    ('company', 'makes', 'game'): (torch.arange(n_nodes)[torch.randperm(n_nodes)], torch.arange(n_nodes)[torch.randperm(n_nodes)])})

for ntype in G.ntypes:
    v = torch.rand(G.num_nodes(ntype), 100, device = 'cpu').float()
    G.nodes[ntype].data['v'] = v

#40.8 ms ± 503 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
%timeit G.multi_update_all({etype : (fn.copy_u('v', 'v_normalized'), fn.sum('v_normalized', 'v_normalized')) \
                    for etype in  G.canonical_etypes}, cross_reducer = 'mean')

def cross_mean(flist):
    sf = torch.stack(flist, dim=0) 
    return torch.sum(sf, dim=0)/(torch.count_nonzero(sf, dim=0)+1e-6)

#90.2 ms ± 1.01 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
%timeit G.multi_update_all({etype : (fn.copy_u('v', 'v_normalized'), fn.sum('v_normalized', 'v_normalized')) \
                    for etype in  G.canonical_etypes}, cross_reducer = cross_mean)
```

**Comment by fmellomascarenhas** [comment]
@Rhett-Ying I was thinking about it, and IMHO, the standard mean should be  case I proposed. It doesn't seem to make too much sense to me to take the mean considering the absence of edgetype connections that don't exist for a particular node id. If the user wanted the connections to exist, they would create it. And if they wanted the value to be divided by a constant, they could sum and divide by the constant after or before the sum.

The great advantage of having an API that calculates de mean of incoming edges is that it is hard to access count the number of incoming edge types per node id. Please,  let me know your thoughts. Thanks!

**Comment by Rhett-Ying** [comment]
@fmellomascarenhas Thanks for you thoughts. I will think about it and get back to you.
